### PR TITLE
Add typechecker testsuite

### PR DESCRIPTION
**Pull request by xtuc**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -43,6 +43,7 @@ test-whitelisted-spec:
 	$(REPL) $(SPEC_TEST_DIR)/globals.wast
 	$(REPL) $(SPEC_TEST_DIR)/i32.wast
 	$(REPL) $(SPEC_TEST_DIR)/binary.wast
+	$(REPL) $(SPEC_TEST_DIR)/typecheck.wast --debug
```
Comment (xtuc) [inline Makefile]: Remove the `--debug` when it's ready to be merge.

File: packages/repl/src/index.js [hunk 2]
```diff
@@ -12,6 +12,7 @@ const {
 } = require("webassemblyjs/lib/interpreter/kernel/memory");
 const { decode } = require("@webassemblyjs/wasm-parser");
 const t = require("@webassemblyjs/ast");
+const typeCheck = require("@webassemblyjs/validation").stack;
```
Comment (maurobringolf) [inline packages/repl/src/index.js]: @xtuc I hope I didnt misunderstand things but to me it looked like no validation was running in the repl at all? I added it in a way that it is currently only run when using `assert_invalid` with `type mismatch` as expected error. As expected some case with `br` and `return` fails :grin:

File: packages/validation/src/type-checker.js [hunk 3]
```diff
@@ -132,7 +133,7 @@ function checkStacks(expectedStack, actualStack) {
     }
 
     // There are still types left on the resulting stack
-    if (j >= 0) {
+    if (actualStack.slice(0, j + 1).filter(t => t !== POLYMORPHIC).length > 0) {
```
Comment (xtuc) [inline packages/validation/src/type-checker.js]: Could we extract this into a new function? It's difficult to understand to me

Comment (maurobringolf) [reply packages/validation/src/type-checker.js]: Sure, I'll make it more explicit. This is saying "If the stack contains anything else than polymorphic types". The reason is that a polymorphic type can also be `[]` so its valid in that case even though there are still types left on the stack after the function.

File: packages/repl/src/index.js [hunk 4]
```diff
@@ -293,9 +282,19 @@ export function createRepl({ isVerbose, onAssert, onLog, onOk }) {
       _internalInstanceOptions: internalInstanceOptions
     };
 
-    const module = createCompiledModule(moduleNode);
+    if (enableTypeChecking === true) {
+      denormalizeTypeReferences(moduleNode);
```
Comment (maurobringolf) [inline packages/repl/src/index.js]: Ah this is much better, thanks :+1:

File: packages/repl/src/index.js [hunk 5]
```diff
@@ -278,9 +281,20 @@ export function createRepl({ isVerbose, onAssert, onLog, onOk }) {
     const importObject = {
       _internalInstanceOptions: internalInstanceOptions
     };
-    const module = createCompiledModule(moduleNode);
 
-    return new Instance(module, importObject);
+    if (enableTypeChecking === true) {
+      denormalizeTypeReferences(moduleNode);
+
+      const typeErrors = typeCheck(t.program([moduleNode]));
```
Comment (xtuc) [inline packages/repl/src/index.js]: I just noticed that but type checking should be done an a module directly.

- Program can have multiple modules
- Program have no typing rules
- in fact I created the notion of Program

File: packages/repl/src/index.js [hunk 6]
```diff
@@ -278,9 +281,20 @@ export function createRepl({ isVerbose, onAssert, onLog, onOk }) {
...
+
+      if (typeErrors.length > 0) {
+        throw new Error("type mismatch");
+      }
+    }
+
+    const compiledModule = createCompiledModule(moduleNode);
```
Comment (xtuc) [inline packages/repl/src/index.js]: See the implementation of `createCompiledModule`: https://github.com/xtuc/webassemblyjs/blob/master/packages/webassemblyjs/src/compiler/compile/module.js#L40-L43. 

It calls the root validation function which includes the new typechecker.

Why do we need to run the typechecker before again? It seems that there's a bug.

**Comment by xtuc** [comment]
I just noticed that createCompiledModule takes a Program, I made to track/change it https://github.com/xtuc/webassemblyjs/issues/322.

**Pull request merged**